nvflare/private/fed/server/server_runner.py

```diff
@@ -92,57 +96,69 @@ def __init__(self, config: ServerRunnerConfig, job_id: str, engine: ServerEngine
-                        # the job monitor to join.
-                        self.current_wf = None
-
-                    self.log_info(fl_ctx, f"Workflow: {wf.id} finalizing ...")
-                    try:
-                        wf.responder.finalize_run(fl_ctx)
-                    except BaseException as e:
-                        self.log_exception(
-                            fl_ctx, "Error finalizing workflow {}: {}".format(wf.id, secure_format_exception(e))
-                        )
+            with self.engine.new_context() as fl_ctx:
+                wf.responder.control_flow(self.abort_signal, fl_ctx)
+        except BaseException as e:
+            with self.engine.new_context() as fl_ctx:
+                self.log_exception(fl_ctx, "Exception in workflow {}: {}".format(wf.id, secure_format_exception(e)))
+            self.system_panic("Exception in workflow {}: {}".format(wf.id, secure_format_exception(e)), fl_ctx)
+        finally:
+            with self.engine.new_context() as fl_ctx:
+                # do not execute finalize_run() until the wf_lock is acquired
+                with self.wf_lock:
+                    # unset current_wf to prevent message processing
+                    # then we can release the lock - no need to delay message processing
+                    # during finalization!
+                    # Note: WF finalization may take time since it needs to wait for
+                    # the job monitor to join.
+                    self.current_wf = None
+
+                self.log_info(fl_ctx, f"Workflow: {wf.id} finalizing ...\n")
+                try:
+                    wf.responder.finalize_run(fl_ctx)
+                except BaseException as e:
+                    self.log_exception(
+                        fl_ctx, "Error finalizing workflow {}: {}".format(wf.id, secure_format_exception(e))
+                    )
 
-                    self.log_debug(fl_ctx, "firing event EventType.END_WORKFLOW")
-                    self.fire_event(EventType.END_WORKFLOW, fl_ctx)
+                self.log_debug(fl_ctx, "firing event EventType.END_WORKFLOW")
+                self.fire_event(EventType.END_WORKFLOW, fl_ctx)
 
-                # Stopped the server runner from the current responder, not continue the following responders.
-                if self.abort_signal.triggered:
-                    break
-            self.current_wf_index += 1
+    def _get_target_workflow(self) -> list:
+        wf_ids = self.config.target_wf_ids
+        if wf_ids is None:
+            return self.config.workflows
+        else:
+            target_workflows = []
+            for wf_id in wf_ids:
+                target_wfs = [wf for wf in self.config.workflows if wf.id == wf_id]
+                target_workflows.extend(target_wfs)
+        return target_workflows
+
+    def _execute_run(self):
+        target_wf = self._get_target_workflow()
+        for wf in target_wf:
```

Comment:
the previous code using "while self.current_wf_index < len(self.config.workflows):" for a reason.

If you check code: https://github.com/NVIDIA/NVFlare/blob/dev/nvflare/private/fed/server/server_runner.py#L482-L487

You see we will persist the `current_wf_index`.
And when HA resumes, we will only execute the workflows that has not been finished.

Reply:
Please make sure to take HA into consideration, thanks